File: Python/problem_010.py
from timeit import default_timer as timer


def sieve_of_eratosthenes(number):
    # https://en.wikipedia.org/wiki/Sieve_of_Eratosthenes
    # https://www.geeksforgeeks.org/sieve-of-eratosthenes

    # Create a boolean array "prime[0..n]" and initialize
    #  all entries it as true. A value in prime[i] will
    # finally be false if i is Not a prime, else true.
    prime = [True for i in range(number + 1)]
    p = 2
    primes = []

    while p * p <= number:
        # If prime[p] is not changed, then it is a prime
        if prime[p]:
            # Update all multiples of p
            for i in range(p * p, number + 1, p):
                prime[i] = False
        p += 1
    # Add up all prime numbers
    for p in range(2, number + 1):
        if prime[p]:
            primes.append(p)

    return primes


def problem_010():
    total_sum, target = 0, 2000000
    # Get all primes below target
    primes = sieve_of_eratosthenes(target - 1)

    for prime in primes:
        total_sum += prime

    return total_sum


if __name__ == '__main__':
    start = timer()
    answer = problem_010()
    elapsed_time = (timer() - start) * 1000
    print(f"Found {answer} in {elapsed_time} ms")

File: Java/src/main/java/nl/davydehaas/projecteuler/Problem010.java
package nl.davydehaas.projecteuler;

import static nl.davydehaas.projecteuler.util.SolutionTimer.timeSolution;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Problem010 {

    void main() {
        timeSolution(Problem010::solve);
    }

    static Number solve() {
        int target = 2_000_000;

        List<Integer> primes = sieveOfEratosthenes(target - 1);

        return primes.stream()
                .mapToLong(Integer::longValue)
                .sum();
    }

    /// Returns all prime numbers from 1 to `target`
    ///
    /// - [Wikipedia](https://en.wikipedia.org/wiki/Sieve_of_Eratosthenes)
    /// - [Geeks For Geeks](https://www.geeksforgeeks.org/sieve-of-eratosthenes)
    ///
    /// @return [List] of primes as [Integer]
    static List<Integer> sieveOfEratosthenes(int target) {
        // Create a boolean array "isPrime[0..n] and initialize all entries as true.
        // A value in isPrime[i] will be false if i is not a prime, else true.
        boolean[] isPrime = new boolean[target + 1];
        Arrays.fill(isPrime, true);

        int current = 2;
        List<Integer> primes = new ArrayList<>();

        while (current * current <= target) {
            // If isPrime[current] is not changed, then it is a prime
            if (isPrime[current]) {
                // Update all multiples of current
                for (int i = current * current; i < target + 1; i += current) {
                    isPrime[i] = false;
                }
            }
            current += 1;
        }

        // Add up all prime numbers
        for (int i = 2; i < target + 1; i++) {
            if (isPrime[i]) {
                primes.add(i);
            }
        }

        return primes;
    }
}
